### PR TITLE
#164069182 Enable authenticated users to view and edit their profiles

### PR DESCRIPTION
**Pull request by CryceTruly**
**What does this PR do**
This PR Enables users to view and edit their profiles

**Description of tasks to be completed**
- Create the profile component
- Create the edit profile component
- Upload photos to google cloud storage

**How can this be tested??**
- `clone the repo`
- `checkout ft-view-edit-profiles-164069182`
- `npm install && npm start`
- `login with your account credentials`
- `visit /profile`
- `edit profile here /editprofile`

**Any background context you want to provide**
  N/A

#### Applicable screenshots

Screenshot 2019-04-15 at 19 04 03
Screenshot 2019-04-15 at 19 04 29
Screenshot 2019-04-15 at 19 19 55

**What are the relevant pivotal tracker stories**
- [#164069182 ](https://www.pivotaltracker.com/story/show/164069182 )

### CODE REVIEW

**Review by Peace-Apple — APPROVED** [review]
Great work, @CryceTruly checked it out, and good job on sorting code climate issues.

File: src/__tests__/actions/loginAction/__test__/socialAction.test.js [hunk 1]
```diff
@@ -35,4 +35,12 @@ describe("Tests facebook actions", () => {
     store.dispatch(socialAction({ access: "token" }, "facebook"));
     expect(store.getActions()).toEqual([]); 
   });
+
+
+
+
+
```
Comment (joelethan) [inline src/__tests__/actions/loginAction/__test__/socialAction.test.js]: Good work @CryceTruly However, these redundant lines need to be removed to improve the code quality

**Review by joelethan — CHANGES_REQUESTED** [review]

File: src/__tests__/actions/profile/profile.test.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    const expectedActions = [{ type: "PROFILE_LOADING" }, {
+      payload: {
+        bio: "samyy",
+        image: "https://m0yfsbokq47e8jxhhbos.jpg",
+        username: "testname",
+      },
+      type: PROFILE_LOADED,
+    }];
+
+
```
Comment (joelethan) [inline src/__tests__/actions/profile/profile.test.js]: @CryceTruly Please remove these redundant blank lines.

File: src/__tests__/actions/loginAction/__test__/socialAction.test.js (hunk 1 above)
Comment (CryceTruly) [reply src/__tests__/actions/loginAction/__test__/socialAction.test.js]: Thanks, @joelethan, thanks for pointing out this eslint issue. I have refactored it.

File: src/__tests__/actions/profile/profile.test.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+      type: PROFILE_LOADED,
+    }];
+    const store = mockStore();
+    return store
+      .dispatch(loadProfile())
+      .then(() => expect(store.getActions())
+        .toEqual(expectedActions));
+  });
+
+
```
Comment (joelethan) [inline src/__tests__/actions/profile/profile.test.js]: @CryceTruly This blank line needs to be worked on as well.

File: src/actions/profile/profileActions.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    body,
+
+  })
+    .then(response => response.json())
+    .then((data) => {
+      if (data.profile) {
+        try {
+          history.push("/profile");
+        } catch (e) {
+
```
Comment (joelethan) [inline src/actions/profile/profileActions.js]: @CryceTruly line 50. Why don't we have what should happen in the event that the error is caught?

**Review by joelethan — CHANGES_REQUESTED** [review]

File: src/__tests__/actions/profile/profile.test.js (hunk 2 above)
Comment (CryceTruly) [reply src/__tests__/actions/profile/profile.test.js]: Thanks @joelethan ,am done working on this

File: src/__tests__/actions/profile/profile.test.js (hunk 3 above)
Comment (CryceTruly) [reply src/__tests__/actions/profile/profile.test.js]: Thanks @joelethan ,am done working on this too.

Comment (joelethan) [reply src/__tests__/actions/profile/profile.test.js]: @CryceTruly Am seeing this is still there.

File: src/__tests__/components/profile/EditProfile.test.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    const e = {
+      preventDefault: () => {
+      },
+      target: {
+        name: "hey", value: "How are you",
+      },
+    };
+    wrapper.instance().onSubmit(e);
+  });
+});
```
Comment (joelethan) [inline src/__tests__/components/profile/EditProfile.test.js]: This is missing a blankline at the end.

**Review by joelethan — CHANGES_REQUESTED** [review]

**Review by joelethan — APPROVED** [review]

**Review by wasibani-roy — DISMISSED** [review]
Good work @CryceTruly i see all feedback implemented

File: src/__tests__/actions/profile/profile.test.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    url = `${API_HOST_URL}/profiles/${username}`;
+    store.clearActions();
+  });
+  afterEach(() => {
+    username = "";
+    url = "";
+    fetchMock.restore();
+  });
+
+  it("tests if can load  a user profile", () => {
```
Comment (AmosWels) [inline src/__tests__/actions/profile/profile.test.js]: check the grammar for title of your `it` block

File: src/__tests__/actions/profile/profile.test.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        username: "testname",
+      },
+      type: PROFILE_LOADED,
+    }];
+    return store.dispatch(loadProfile())
+      .then(() => expect(store.getActions())
+        .toEqual(expectedActions));
+  });
+
+  it("test should fail to update profile", () => {
```
Comment (AmosWels) [inline src/__tests__/actions/profile/profile.test.js]: same here. Check elsewhere as well

File: src/__tests__/components/profile/EditProfile.test.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      target: {
+        files: [new File(["../../../assets/download.jpeg"], "sample.png")],
+      },
+    };
+
+    wrapper.instance().handleChange(e);
+    const image = wrapper.instance().state.image;
+    wrapper.instance().handleUpload(image);
+    expect(wrapper.state().image.name).toEqual("sample.png");
+  });
```
Comment (AmosWels) [inline src/__tests__/components/profile/EditProfile.test.js:31]: leave a blank line here

File: src/__tests__/components/profile/EditProfile.test.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    wrapper = mount(<EditProfile {...props} />);
+  });
+
+  it("should handleChange", () => {
+    const e = {
+      target: {
+        files: [new File(["../../../assets/download.jpeg"], "sample.png")],
+      },
+    };
+
```
Comment (AmosWels) [inline src/__tests__/components/profile/EditProfile.test.js]: remove this blank line

File: src/__tests__/components/profile/Profile.test.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+import { Provider } from "react-redux";
+import store from "../../../store/index";
+import { loadProfile } from "../../../actions/profile/profileActions";
+import Profile from "../../../components/profile/profile";
+
+describe("profile component tests", () => {
+  it("should not regress", () => {
+    const wrapper = shallow(<Profile />);
+    expect(wrapper).toMatchSnapshot();
+  });
```
Comment (AmosWels) [inline src/__tests__/components/profile/Profile.test.js:13]: add blank line here

File: src/__tests__/components/profile/Profile.test.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
+import React from "react";
+import { shallow, mount } from "enzyme";
+import { Provider } from "react-redux";
+import store from "../../../store/index";
+import { loadProfile } from "../../../actions/profile/profileActions";
+import Profile from "../../../components/profile/profile";
+
+describe("profile component tests", () => {
+  it("should not regress", () => {
```
Comment (AmosWels) [inline src/__tests__/components/profile/Profile.test.js:10]: what does `regress` mean here

File: src/__tests__/reducers/profileReducer.test.js [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        image: "testimage",
+      },
+      isLoading: null,
+      msg: null,
+      isUpdating: null,
+    };
+    expect(profileReducer(initialState, successAction)).toEqual(successState);
+  });
+
+  it("should it fail to load a users profile", () => {
```
Comment (AmosWels) [inline src/__tests__/reducers/profileReducer.test.js]: correct the grammar here

File: src/__tests__/reducers/profileReducer.test.js [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+      payload: "something went wrong",
+    };
+    const newState = {
+      profile: null,
+      isLoading: null,
+      msg: "something went wrong",
+      isUpdating: null,
+    };
+    expect(profileReducer(initialState, dispatchedAction)).toEqual(newState);
+  });
```
Comment (AmosWels) [inline src/__tests__/reducers/profileReducer.test.js:65]: add blank line here

File: src/__tests__/reducers/profileReducer.test.js [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+        email: "crycetruly@gmail.com",
+        bio: "this is the bio",
+        image: "testimage",
+      },
+      isLoading: null,
+      msg: null,
+      isUpdating: null,
+    };
+    expect(profileReducer(initialState, dispachedAction)).toEqual(newProfileState);
+  });
```
Comment (AmosWels) [inline src/__tests__/reducers/profileReducer.test.js:133]: add new blank line

**Review by AmosWels — CHANGES_REQUESTED** [review]
Goodwork @CryceTruly , add the `package-lock.json` file to your gitignore. 

Also check that your tests' file naming format is consistent.

File: src/__tests__/actions/profile/profile.test.js (hunk 6 above)
Comment (CryceTruly) [reply src/__tests__/actions/profile/profile.test.js]: Thanks @AmosWels, I have made all file names to camelcase just like the other code files

File: src/__tests__/components/profile/Profile.test.js (hunk 11 above)
Comment (CryceTruly) [reply src/__tests__/components/profile/Profile.test.js:10]: Good question @AmosWels , It basically confirms that recent program changes or code changes have not adversely affected the functionality of this feature.

File: src/actions/profile/profileActions.js (hunk 4 above)
Comment (CryceTruly) [reply src/actions/profile/profileActions.js]: Hey @joelethan,   I have rewritten the logic using conditional if statements.

File: src/__tests__/components/profile/EditProfile.test.js (hunk 5 above)
Comment (CryceTruly) [reply src/__tests__/components/profile/EditProfile.test.js]: Have added that,

File: src/__tests__/actions/profile/profile.test.js (hunk 7 above)
Comment (CryceTruly) [reply src/__tests__/actions/profile/profile.test.js]: Okay @AmosWels

**Review by AmosWels — APPROVED** [review]

**Pull request merged**